File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        
        int n = nums.length;
        List<Integer> lis = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            int cur = nums[i];

            int idx = bs(cur, lis);
            if (idx == lis.size()) lis.add(cur);
            lis.set(idx, cur);
        }

        return lis.size();
    }

    private int bs(int t, List<Integer> list) {
        int l = 0, r = list.size();

        while (l < r) {
            int m = l + ((r - l) >> 2);

            if (list.get(m) < t) l = m + 1;
            else r = m;
        }

        return l;
    }
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
#include <vector>
using namespace std;

class Solution {
private:
    vector<vector<int>> dp;
    vector<int> arr;
    int n;

public:
    int lengthOfLIS(vector<int>& nums) {
        arr = nums;
        n = arr.size();
        dp.assign(n, vector<int>(n + 1, -1));

        return dfs(0, -1);
    }

private:
    int dfs(int cur, int prev) {
        if (cur >= n) {
            return 0;
        }

        if (dp[cur][prev + 1] != -1) {
            return dp[cur][prev + 1];
        }

        int skip = dfs(cur + 1, prev);

        int take = 0;
        if (prev == -1 || arr[prev] < arr[cur]) {
            take = 1 + dfs(cur + 1, cur);
        }

        return dp[cur][prev + 1] = max(skip, take);
    }
};